### PR TITLE
refactor: introduce the packetmuxer layer

### PR DESCRIPTION
**Pull request by ainghazal**
This is the second commit in the series of incremental refactoring of the current minivpn tree.

In this commit, we introduce the packet muxer, which is the layer just above the networkio. The packet muxer handles and routes data or control packets, and it also handles reset packets.

As dependencies for `packetmuxer`, we're also introducing the `optional` and `session` packages.

Reference issue: #47

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -10,7 +10,7 @@ jobs:
     - name: setup go
       uses: actions/setup-go@v2
       with:
-        go-version: '1.18'
```
Comment (ainghazal) [inline .github/workflows/build.yml:13]: i'm not really sure what's going on here - this particular change should not be in the diff in my opinion - it's already set to 1.20 in main and on this branch.

Comment (bassosimone) [reply .github/workflows/build.yml:13]: I think you need to merge _again_ `ooni/main` into the working branch after https://github.com/ooni/minivpn/commit/d976c5ab4eb66f4f3b3b811ae0ce325a2adaa809 has been merged; you're probably missing this commit in your working branch.

File: internal/packetmuxer/service.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+		select {
+		case packet := <-ws.dataOrControlToMuxer:
+			// serialize the packet
+			rawPacket, err := packet.Bytes()
+			if err != nil {
+				ws.logger.Warnf("packetmuxer: cannot serialize packet: %s", err.Error())
+				continue
+			}
+
+			// While this channel send could possibly block, the [ARCHITECTURE] is
```
Comment (ainghazal) [inline internal/packetmuxer/service.go:145]: i've left this comment here as a marker to consider what was our original intention; I think the whole situation might have changed a bit from where we started drafting the initial architectural stub. Whatever assumptions are valid should probably be documented in the user API, i.e., the tun interface

File: internal/session/datachannelkey.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	"fmt"
+	"sync"
+)
+
+// DataChannelKey represents a pair of key sources that have been negotiated
+// over the control channel, and from which we will derive local and remote
+// keys for encryption and decrption over the data channel. The index refers to
+// the short key_id that is passed in the lower 3 bits if a packet header.
+// The setup of the keys for a given data channel (that is, for every key_id)
+// is made by expanding the keysources using the prf function.
```
Comment (bassosimone) [inline internal/session/datachannelkey.go:14]:
```suggestion
// is made by expanding the keysources using the prf function.
//
```

File: internal/session/manager.go [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+	logger               model.Logger
+	mu                   sync.Mutex
+	negState             SessionNegotiationState
+	remoteSessionID      optional.Value[model.SessionID]
+	tunnelInfo           model.TunnelInfo
+
+	// Ready is a channel where we signal that we can start accepting data, because we've
+	// successfully generated key material for the data channel.
+	// TODO(ainghazal): find a better way?
+	Ready chan (any)
```
Comment (bassosimone) [inline internal/session/manager.go:91]:
```suggestion
	Ready chan any
```

File: internal/session/manager.go [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+	defer m.mu.Unlock()
+	m.mu.Lock()
+	return m.localDataPacketIDLocked()
+}
+
+// localDataPacketIDLocked returns an unique Packet ID for the Data Channel. It
+// increments the counter for the local data packet ID.
+func (m *Manager) localDataPacketIDLocked() (model.PacketID, error) {
+	pid := m.localDataPacketID
+	if pid == math.MaxUint32 {
```
Comment (bassosimone) [inline internal/session/manager.go:225]:
```suggestion
	if pid >= math.MaxUint32 {
```

File: internal/session/manager.go [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+	}
+	m.localDataPacketID++
+	return pid, nil
+}
+
+// localControlPacketIDLocked returns an unique Packet ID for the Control Channel. It
+// increments the counter for the local control packet ID.
+func (m *Manager) localControlPacketIDLocked() (model.PacketID, error) {
+	pid := m.localControlPacketID
+	if pid == math.MaxUint32 {
```
Comment (bassosimone) [inline internal/session/manager.go:237]:
```suggestion
	if pid >= math.MaxUint32 {
```

**Review by bassosimone — APPROVED** [review]
🐳

File: internal/session/manager.go (hunk 5 above)
Comment (ainghazal) [reply internal/session/manager.go:225]: I think this is not correct right? PacketID is a uint32 so it should not overflow.
This is a very improbable error, but the only way of returning it meaningfully is to error on the last usable PacketID, I think.

File: internal/session/manager.go (hunk 6 above)
Comment (ainghazal) [reply internal/session/manager.go:237]: same as above.

File: internal/session/manager.go (hunk 5 above)
Comment (bassosimone) [reply internal/session/manager.go:225]: I tend to like expressing this kind of checks in a more defensive way which makes them more robust to refactoring like "let's change this to int64 now", but feel free to ignore if you don't like it.

**Pull request merged**